dulwich/porcelain.py

```diff
@@ -172,16 +172,29 @@ def open_repo_closing(path_or_repo):
 
 
 def path_to_tree_path(repopath, path):
-    """Convert a path to a path usable in e.g. an index.
+    """Convert a path to a path usable in an index, e.g. bytes and relative to
+    the repository root.
 
-    :param repo: Repository
-    :param path: A path
+    :param repopath: Repository path, absolute or relative to the cwd
+    :param path: A path, absolute or relative to either the cwd or repopath
     :return: A path formatted for use in e.g. an index
     """
-    os.path.relpath(path, repopath)
+    if not isinstance(path, bytes):
+        path = path.encode(sys.getfilesystemencoding())
+    if not isinstance(repopath, bytes):
+        repopath = repopath.encode(sys.getfilesystemencoding())
     if os.path.sep != '/':
-        path = path.replace(os.path.sep, '/')
-    return path.encode(sys.getfilesystemencoding())
+        path = path.replace(
+            os.path.sep.encode(sys.getfilesystemencoding()), b'/')
+        repopath = repopath.replace(
+            os.path.sep.encode(sys.getfilesystemencoding()), b'/')
+    if os.path.exists((repopath + b'/' + path).replace(b'//', b'/')):
```

Comment:
See above - this makes the behaviour of the function unpredictable, relative paths should only ever be interpreted in one way.